File: Trees/Symmetric tree.py
 #Recursive apprach
def isMirror(root1,root2):
    if root1==None and root2==None:
        return True
    if root==None or root2==None:
        return False
    
    return ((root1.data==root2.data) and isMirror(root1.left,root2.right) and isMirror(root1.right,root2.left))
def isSymmetric(root):
    
    return isMirror(root,root)
  ###################################################
  
  #Iterative approach
  
      def isSymmetric(self, root):
        if root==None:
            return True
        queue = [root.left,root.right]
        while queue!=[]:
            temp_left = queue.pop(0)
            temp_right = queue.pop(0)
            
            if temp_left==None and temp_right == None:
                continue
            if temp_left==None or temp_right == None:
                return False
            if temp_left.val!=temp_right.val:
                return False
            queue.append(temp_left.left)
            queue.append(temp_right.right)
            queue.append(temp_left.right)
            queue.append(temp_right.left)
        return True
                     

File: Trees/Symmetric tree.java
 #Recursive apprach
//return true/false if the is Symmetric or not
class GfG
{
    // return true/false denoting whether the tree is Symmetric or not
    public static boolean isSymmetric(Node root)
    {
        return isSymmetricHelper(root,root);
    }
    
    public static boolean isSymmetricHelper(Node root1,Node root2){
        if(root1==null && root2==null)return true;
        if(root1==null || root2==null)return false;
        return root1.data==root2.data && isSymmetricHelper(root1.left,root2.right)&& isSymmetricHelper(root1.right,root2.left);
    }
}
  ###################################################
  
  #Iterative approach
  
class GfG
{
    // return true/false denoting whether the tree is Symmetric or not
    public static boolean isSymmetric(Node root)
    {
        if(root==null)return true;
        Queue<Node> q = new LinkedList<Node>();
        q.add(root.left);
        q.add(root.right);
        while(!q.isEmpty()){
            Node temp_left  = q.remove();
            Node temp_right = q.remove();
            if(temp_left==null && temp_right==null)continue;
            if(temp_left==null || temp_right==null)return false;
            if(temp_left.data!=temp_right.data)return false;
            q.add(temp_left.left);
            q.add(temp_right.right);
            q.add(temp_left.right);
            q.add(temp_right.left);
        }
        return true;
    }
    
    
}
                     
